### DISCUSSION TITLE
Roc on microcontroller?

### DISCUSSION THREAD

**Discussion by elrom** [roc-streaming/roc-toolkit#478] [Discuss]
Has anyone tried running roc on a microcontroller on bare metal? Authors - any idea how hard it would to get that running?

**Comment by baranovmv** [comment]
Hi elrom,

Thank you for your interest. At the beginning we planned to keep the
codebase self-sufficient to be able to run on bare-metal embedded or some
kind of RTOS. However, in the course of development roc adopted an
asynchronous approach for multithreading programming, and uses libuv
internally. Which is the main blocker for compiling for bare metal embedded
platforms.

In my personal opinion, there are no some isolated spots in the codebase
which potentially could be somehow tweaked to be able to run with say lwip
instead of posix compliant  standard library.

If you are really keen on this task, I suppose you'd better take the
processing pipeline code of the roc and redo the environment. If you wish,
we could discuss this approach in more detail here

On Mon, 29 Aug 2022 at 01:23, elrom ***@***.***> wrote:

> Has anyone tried running roc on a microcontroller on bare metal? Authors -
> any idea how hard it would to get that running?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/roc-streaming/roc-toolkit/discussions/478>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAO4V53KP7VY2O2QTCZTVVLV3PKD7ANCNFSM573WZMFQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 

Best regards,
Mikhail Baranov,
+7(926)559-44-01
***@***.***

**Comment by gavv** [comment]
We still have pipeline and I/O layers very well isolated (they communicated via tasks and queues), as well as platform-specific code is isolated in target directories.

I think running on bare metal fits current deign well (i.e. the whole pipeline may be reused), but of course this will require quite a lot of work (reimplementing network layer for example).

**Comment by gavv** [comment]
See also https://github.com/roc-streaming/roc-toolkit/discussions/496
